[DP-154] As a reader, I want to see the live table of contents for CWKW

chanmioh opened the pull request:
# Changelog
Connected backend query for getting chapters of an `EditedCollection` and displaying them in a table of contents on the homepage.

## Desktop
![image](https://user-images.githubusercontent.com/40529597/197891440-8fd580dc-3c1f-4e24-ab37-2697ceda30d0.png)

![image](https://user-images.githubusercontent.com/40529597/197891465-02cca7ac-3181-41e9-9505-479348229126.png)

### When a chapter is selected, only its **direct** descendants are viewable. 
![image](https://user-images.githubusercontent.com/40529597/197891534-ad8c5ac4-d253-4216-97ff-73eedd26b94e.png)

## Mobile
![image](https://user-images.githubusercontent.com/40529597/197891571-6b07078a-402c-4330-ae24-933a9749eddb.png)

![image](https://user-images.githubusercontent.com/40529597/197891596-5b07d9f4-c811-4372-89f9-d65299628b1f.png)

![image](https://user-images.githubusercontent.com/40529597/197891625-9a3b5204-2e82-4028-addd-1aea4bbdaa26.png)

## Files Changed
- Created `edited-collections` folder containing files for any collection, not just CWKW.
- `queries.graphql`: Added `EditedCollection` query for getting a collection's chapters and each chapter's fields.
- `toc.page.tsx`: 
     - Fetches all the chapters of a collection and converts it into a nested data structure according to its `indexInParent` and `path`.
     - Renders all chapters according to their nested structure.
- `sidebar.tsx`: Modified to utilize a `Dialog` for the TOC.
- `cwkw-layout.tsx`: Modified to contain the CWKW banner.
- `dialog-context.tsx`: Added a dialog context, this is used for detecting if the TOC is visible or not. When the TOC is visible, elements on the page are shifted to the right to give space for the TOC.

hazelyn11 commented on website/src/components/toc.page.tsx:
```diff
@@ -1,34 +1,109 @@
-    </ol>
-  </>
-)
+interface Chapters {
```
Once we merge the changes with the additional type of Credit chapter, there might have to be an additional type here

hazelyn11 replied:
Which I guess also brings up a different question I have: Are the different types of chapters mostly for organization within the spreadsheet and database? Since all chapters have the option to display a document, for instance, and the same fields, how relevant is distinguishing between them for display purposes? Not sure if that question makes sense but just some thoughts looking at the code brought up

chanmioh replied:
I think it's mostly for organization, so mainly a really semantic distinction, though they have the same structure!

hazelyn11 commented on website/src/pages/edited-collections/edited-collection-context.tsx:
```diff
@@ -0,0 +1,165 @@
+        // Check if the current chapter's parent leaf matches the last pushed chapter's leaf.
+        // If it doesn't, the last pushed chapter is not the parent and needs to be popped.
+        // Continue through the stack until the parent of this chapter is found.
+        while (parentChapterLeaf !== lastPushed?.leaf) {
```
Such a small thing but is there a way to consolidate the logic of the previous two lines of code with this while loop, where you're checking if chapters in the stack are the parent? If I'm reading correctly it looks like right now you check the first element in the stack, and then have a while loop

chanmioh replied:
Oh do you mean like remove the `if (curr.indexInParent === 1) {` and utilize the while loop instead?

chanmioh replied:
Actually I just removed most of this code and rewrote it with a `HashMap` instead so it's quicker. If you have time could you look over it again? Tysm!!

hazelyn11 commented on website/src/pages/edited-collections/edited-collection-context.tsx:
```diff
@@ -0,0 +1,165 @@
+import { DialogStateReturn, useDialogState } from "reakit/Dialog"
+import * as Dailp from "src/graphql/dailp"
+import { useRouteParams } from "src/renderer/PageShell"
+
```
All the changes you made in this file are overall super well commented!! As someone who's not really familiar with the language I was still able to follow all the changes you're making.

chanmioh replied:
ty for taking a look at my long dreadful code haha!! i'm glad it wasn't too hard to understand, i made sure to put as many comments as i could so it wasn't too confusing

hazelyn11 commented on website/src/pages/edited-collections/edited-collection.page.tsx:
```diff
@@ -0,0 +1,63 @@
+          </header>
+
+          <h3>
+            A digital collection presented by{" "}
```
We've talked about this in a few meetings already lol but love the landing page and love the linking back to DAILP!!

chanmioh replied:
Tysm!! I'm glad y'all are liking it :)

GracefulLemming reviewed (approved):
As a general comment, I like the way you handled the mobile and desktop displays for the CWKW TOC! Lets go ahead and get ready to merge this

hazelyn11 reviewed (commented):
Like we've talked about in various meetings, love what you've done here and the thought put into both desktop and mobile!! Your code was also very well commented and easy to follow. I just had one or two thoughts/questions which you'll see in my comments.

chanmioh commented:
# Changelog
- Added new chapter type, `credit` to the `Chapters` interface.
- Removed Stack structure and replaced with a `HashMap` for assigning subchapters to their respective parent chapters. Hopefully it should work the same, but it should be much faster now.
